File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def inorder_dfs(node, lower, upper):
            if not node:
                return True
            
            if not (lower < node.val < upper):
                return False
            
            return inorder_dfs(node.left, lower, node.val) and inorder_dfs(node.right, node.val, upper)
        return inorder_dfs(root, -float('inf'), float('inf'))

"""
=========================================================
KEY LEARNINGS: Validate Binary Search Tree (LeetCode 98)
=========================================================

CORE CONCEPTS:
1. Top-Down Bounds: A node isn't just compared to its parent; 
   it is constrained by EVERY ancestor above it. Pass `lower` 
   and `upper` bounds down the recursive calls.
2. Short-Circuiting: Use `return dfs(left) and dfs(right)` to 
   instantly kill the traversal the moment a `False` is found, 
   preventing you from wasting time on the rest of the tree.

GUIDING HINTS:
- Left Child updates the UPPER bound: `dfs(node.left, lower, node.val)`
- Right Child updates the LOWER bound: `dfs(node.right, node.val, upper)`
- Python handles infinity with `float('inf')` and `float('-inf')`.
=========================================================
"""

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int maxSum = Integer.MIN_VALUE; //Keep a maximum sum counter
   
    public int dfs(TreeNode node) { 
        if (node == null) return 0;

        int left_val = Math.max(0, dfs(node.left)); // If left_val is going to be negative, no point adding it
        int right_val = Math.max(0, dfs(node.right));
        
        int currValue = node.val;
        this.maxSum = Math.max(maxSum, currValue + left_val + right_val);

        return currValue + Math.max(left_val, right_val);

    }

    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        this.maxSum = root.val; // Initialise with the root value first
        
        dfs(root);
        // Run DFS from the root, finding the maximum diameter of the tree and comparing the sum to always get the maximum
        return this.maxSum;
    }
}

/**
=========================================================
KEY LEARNINGS: Binary Tree Maximum Path Sum (LeetCode 124)
=========================================================

CORE CONCEPTS:
1. Split vs. Straight Line: 
   - A path can only "split" once (forming the peak of a mountain). 
     We check this sum to update our global tracker: `node.val + left + right`.
   - When returning to a parent, we must choose a single straight line: 
     `node.val + max(left, right)`.
2. Active Pruning: Never block a negative node from being searched. 
   Instead, prune negative branch contributions on the way back up using 
   `Math.max(0, dfs(child))`.

GUIDING HINTS:
- Initialize the global maximum tracker to `Integer.MIN_VALUE` (Java) or 
  `float('-inf')` (Python) so a tree of purely negative values updates correctly.
- Post-Order Traversal: You must calculate your children's values fully 
  before you can make a decision about your own node.
- Time: O(N) because every node is visited exactly once.
- Space: O(H) call stack overhead.
=========================================================
**/
